
import "proto/Config.proto";

package Api.Proto;
option optimize_for = LITE_RUNTIME;

// -----------------------------------------------------------
// requests send by client
// -----------------------------------------------------------

enum Actions {
	kActNone = 0;
	kActInsert = 1;
	kActUpdate = 2;
}

message Update {
	optional Actions action = 1;
	optional int64 id = 2;
	optional int32 storage_idx = 3;
	optional string storage_name = 4;
	optional bytes buffer = 17;
}

enum ClientRequests {
	kCqQuery = 1;
	kCqShortTransact = 32;

	kCqCreateStorage = 256;
	kCqCreateView = 257;
	kCqUnlinkStorage = 258;
	kCqUnlinkView = 259;

	kCqUploadExtern = 260;
	kCqDownloadExtern = 261;
};

message CqQuery {
	optional string view_name = 2;
	repeated bytes keys = 3;
	optional bytes from_key = 4;
	optional bytes to_key = 5;
	optional uint32 limit = 6;
}

message CqShortTransact {
	repeated Update updates = 2;
};

message CqCreateStorage {
	required Db.Proto.StorageConfig config = 1;
}
message CqCreateView {
	required Db.Proto.ViewConfig config = 1;
}

message CqUnlinkStorage {
	required string identifier = 1;
}
message CqUnlinkView {
	required string identifier = 1;
}

message CqUploadExtern {
	required string file_name = 1;
	required bytes buffer = 2;
}
message CqDownloadExtern {
	required string file_name = 1;
}

// -----------------------------------------------------------
// responses send by server
// -----------------------------------------------------------

enum ServerResponses {
	kSrFin = 1;
	kSrRows = 2;
	kSrBlob = 3;
};

enum EnumErrors {
	kErrNone = 0;
	kErrIllegalRequest = 1;
	
	kErrIllegalStorage = 128;
	kErrIllegalView = 129;
}

message SrFin {
	optional bool success = 1;
	optional EnumErrors err_code = 2;
	optional string err_msg = 3;
}

message SrRows {
	repeated bytes row_data = 1;
}

message SrBlob {
	required bytes buffer = 1;
}

